### PR TITLE
Adds support for custom endpoints

### LINKED ISSUE #9788: [FEATURE REQ] eventhubs: Be able to specify a hostname to connect to other than the namespace's FQDN.
In a case where you need to use custom DNS name for a private EventHub endpoint and do not have the ability to setup up DNS forwarding from on-premise to Azure DNS you need to split the Endpoint namespace name from the actual hostname to connect to.

Ideally something like a connection string parameter `customHostName`, where you could specify a host to connect to.

To give an example. Host `myprivateeventhub.servicebus.windows.net` resolves within Azure VNET to `192.168.0.1` (private ip), outside resolves to `35.21.22.45` (public ip). Once you connect to it from on-prem via an ExpressRoute - if you don't have DNS forwarding on-prem it will resolve in a public IP, and you'll go over the internet rather than ER. If you then create a split-horizon DNS zone, say `privateendpoints.company.com` and add an entry there `myprivateeventhub.privateendpoints.company.com` pointing to `192.168.0.1` - you can circumvent the problem of not being able to forward DNS. If you change that in Endpoint connection string now - lib will assume that the namespace name is also your custom FQDN and that, obviously, won't work.

### PR DESCRIPTION
**Pull request by conniey**
* Updates azure-core-amqp to support custom endpoints.
* Exposes customEndpointAddress in Event Hubs.

Fixes #9788

Related to #18996

### CODE REVIEW

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubClientBuilder.java [hunk 1]
```diff
@@ -662,8 +693,13 @@ private ConnectionOptions getConnectionOptions() {
             ? verifyMode
             : SslDomain.VerifyMode.VERIFY_PEER_NAME;
 
+        final String hostname = customEndpointAddress != null ? customEndpointAddress.getHost() : null;
+        final Integer port = customEndpointAddress != null && customEndpointAddress.getPort() != -1
+            ? customEndpointAddress.getPort()
+            : null;
```
Comment (alzimmermsft) [inline sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubClientBuilder.java:699]: Should we just turn this into a single `if/else`

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorConnection.java [hunk 2]
```diff
@@ -97,10 +97,8 @@ public ReactorConnection(String connectionId, ConnectionOptions connectionOption
         this.tokenManagerProvider = Objects.requireNonNull(tokenManagerProvider,
             "'tokenManagerProvider' cannot be null.");
         this.messageSerializer = messageSerializer;
-        this.handler = handlerProvider.createConnectionHandler(connectionId,
-            connectionOptions.getFullyQualifiedNamespace(), connectionOptions.getTransportType(),
-            connectionOptions.getProxyOptions(), product, clientVersion, connectionOptions.getSslVerifyMode(),
-            connectionOptions.getClientOptions());
+        this.handler = handlerProvider.createConnectionHandler(connectionId, product, clientVersion,
```
Comment (hemanttanwar) [inline sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorConnection.java:100]: Nice, number of argument are reduced.

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorHandlerProvider.java [hunk 3]
```diff
@@ -31,44 +29,57 @@
...
-            SslDomain.VerifyMode verifyMode, ClientOptions clientOptions) {
-        switch (transportType) {
+    public ConnectionHandler createConnectionHandler(String connectionId, String productName, String clientVersion,
+        ConnectionOptions options) {
+        Objects.requireNonNull(connectionId, "'connectionId' cannot be null.");
+        Objects.requireNonNull(options, "'options' cannot be null.");
+        Objects.requireNonNull(productName, "'productName' cannot be null.");
+        Objects.requireNonNull(clientVersion, "'clientVersion' cannot be null.");
+
+        switch (options.getTransportType()) {
```
[the remark this answers is not in the document]
Comment (hemanttanwar) [reply sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorHandlerProvider.java:56]: Yes it is required not null in `ConnectionOptions`, so this should be fine.

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorHandlerProvider.java [hunk 4]
```diff
@@ -31,59 +29,75 @@
...
-                    return new WebSocketsProxyConnectionHandler(connectionId, hostname, proxyOptions, product,
-                        clientVersion, verifyMode, clientOptions);
-                } else if (WebSocketsProxyConnectionHandler.shouldUseProxy(hostname)) {
-                    logger.info("System default proxy configured for hostname '{}'. Using proxy.", hostname);
-                    return new WebSocketsProxyConnectionHandler(connectionId, hostname,
-                        ProxyOptions.SYSTEM_DEFAULTS, product, clientVersion, verifyMode, clientOptions);
+                    // options.getPort() and options.getHostname() doesn't matter here because it will be overridden by
+                    // the proxy's port and hostname.
+                    return new WebSocketsProxyConnectionHandler(connectionId, options.getHostname(), options.getPort(),
+                        proxyOptions, productName, clientVersion, options.getSslVerifyMode(),
```
Comment (hemanttanwar) [inline sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorHandlerProvider.java:66]: If  `options.getPort() and options.getHostname() doesn't matter` : Should we have another constructor of `WebSocketsProxyConnectionHandler` which does not require these as a parameter. And we do not need to pass this ?

File: sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorHandlerProvider.java [hunk 5]
```diff
@@ -31,44 +29,57 @@
      * generated handlers.
      *
      * @param provider The provider that creates and manages {@link Reactor} instances.
+     *
+     * @throws NullPointerException If {@code provider} is {@code null}.
      */
     public ReactorHandlerProvider(ReactorProvider provider) {
-        this.provider = provider;
+        this.provider = Objects.requireNonNull(provider, "'provider' cannot be null.");
```
[the remark this answers is not in the document]
Comment (srnagar) [reply sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorHandlerProvider.java:36]: This is internal and if this happens, it's a bug and the user should ideally not have seen this before.

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubClientBuilder.java [hunk 6]
```diff
@@ -267,6 +270,34 @@ public EventHubClientBuilder configuration(Configuration configuration) {
...
+     * an intermediary. For example: {@literal https://my.custom.endpoint.com:55300}.
+     * <p>
+     * If no port is specified, the default port for the {@link #transportType(AmqpTransportType) transport type} is
+     * used.
+     *
+     * @param customEndpointAddress The custom endpoint address.
+     * @return The updated {@link EventHubClientBuilder} object.
+     * @throws IllegalArgumentException if {@code customEndpointAddress} cannot be parsed into a valid {@link URL}.
+     */
+    public EventHubClientBuilder customEndpointAddress(String customEndpointAddress) {
```
Comment (srnagar) [inline sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubClientBuilder.java:285]: The HTTP client builders just use `endpoint` not `endpointAddress`. So, should we use similar terminology here? `customEndpoint()`?

**Review by srnagar — APPROVED** [review]
LGTM!

**Review by alzimmermsft — APPROVED** [review]

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubClientBuilder.java (hunk 1 above)
Comment (conniey) [reply sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubClientBuilder.java:699]: There is another `&& customEndpointAddress().getport()` check here, that's why I didn't.

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubClientBuilder.java (hunk 6 above)
Comment (conniey) [reply sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubClientBuilder.java:285]: https://github.com/Azure/azure-sdk/issues/1367
https://gist.github.com/jsquire/f0948c9a5ffe2741b03d312dcf8d899e#gistcomment-3345232

Looks like the board decided on customEndpointAddress

File: sdk/eventhubs/azure-messaging-eventhubs/CHANGELOG.md [hunk 7]
```diff
@@ -1,7 +1,9 @@
 # Release History
 
-## 5.5.0-beta.1 (Unreleased)
+## 5.5.0 (2021-02-15)
```
Comment (alzimmermsft) [inline sdk/eventhubs/azure-messaging-eventhubs/CHANGELOG.md:3]: nit: This needs to be changed back to 5.5.0-beta.1 to resolve the CI issue

**Pull request merged**